FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -0,0 +1,137 @@
+//
+//  CoreDataFeedStore.swift
+//  FeedStoreChallenge
+//
+//  Created by Oliver Jordy Pérez Escamilla on 23/01/21.
+//  Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import Foundation
+import CoreData
+
+public final class CoreDataFeedStore: FeedStore {
+	
+	// MARK: Properties
+	lazy var context: NSManagedObjectContext = {
+		Self.coreDataStack.context
+	}()
+	private static var coreDataStack: CoreDataStack = CoreDataStack(modelName: "FeedStoreModel")
+	
+	// MARK: Initializers
+	public init(coreDataStack: CoreDataStack) {
+		Self.coreDataStack = coreDataStack
+	}
+	
+	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
+		context.perform { [unowned context] in
+			
+			do {
+				guard let cache = try self.fetchCache() else { return completion(nil) }
+				context.delete(cache)
+				
+				completion(nil)
+				
+			} catch {
+				return completion(error)
+			}
+			
+			Self.coreDataStack.saveContext()
+		}
+	}
+	
+	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
+		context.perform { [unowned self] in
+			
+			do {
+				if let cache = try self.fetchCache() {
+					self.context.delete(cache)
+				}
+				
+			} catch {
+				return completion(error)
+			}
+			
+			let feedCache = FeedCache(context: self.context)
+			let feedImages = transformToManagedRepresentation(feed)
+			
+			feedCache.feedImages = NSOrderedSet(array: feedImages)
+			feedCache.timestamp = timestamp
+			
+			do {
+				try self.context.save()
+			} catch {
+				completion(error)
+			}
+			completion(nil)
+		}
+		
+	}
+	
+	public func retrieve(completion: @escaping RetrievalCompletion) {
+		do {
+			let cache = try fetchCache()
+			
+			let localFeedResult = cache.map {
+				(feed: $0.feedImages, timestamp: $0.timestamp)
+			}
+			
+			guard let feedResult = localFeedResult,
+						let timestamp = localFeedResult?.timestamp
+			else { return completion(.empty) }
+			
+			let feed = feedResult.feed?.array as? [FeedImage] ?? []
```

Comment:
This logic can be simplified in one line:

```suggestion
			guard let cache = try fetchCache(), let feed = cache.feedImages?.array as? [FeedImage] else { return completion(.empty) }
```